table/block.h

```diff
@@ -352,6 +362,15 @@ class BlockIter final : public InternalIterator {
   bool key_pinned_;
   // Key is in InternalKey format
   bool key_includes_seq_;
+  // Whether the value is delta encoded. In that case the value is assumed to be
+  // BlockHandle. The first value in each restart interval is the full encoded
+  // BlockHandle; the restart of encoded size part of the BlockHandle. The
+  // offset of delta encoded BlockHandles is computed by adding the size of
+  // previous delta encoded values in the same restart interval to the offset of
+  // the first value in that restart interval.
+  bool value_delta_encoded_;
+  mutable std::string value_buf_;
+  BlockHandle decoded_value_;
```

Comment:
Having so many index block specific logic in BlockIter can make the code harder to understand. Is there a way to generialize it? Is it possible to define an interface like ValueDeltaEncoder (by default null) and call the interface to delta decode the value? Is it possible?

Reply:
It could be static method of helper class as well.

Reply:
How about make another IndexBlockIter, and move all the index related logic to there, and leave those data block only logic behind? The common binary search can be done with a helper class or a subclass. I do worry that we make the already very complicated BlockIter to be even more complicated.

Reply:
update: after the rebase, the logic is moved to IndexBlockIter.